util/ranger/checker.go

```diff
@@ -24,57 +24,83 @@ import (
+	length                   int
+	optPrefixIndexSingleScan bool
 }
 
-func (c *conditionChecker) check(condition expression.Expression) bool {
+func (c *conditionChecker) isFullLengthColumn() bool {
+	return c.length == types.UnspecifiedLength || c.length == c.checkerCol.GetType().GetFlen()
+}
+
+// check returns two values, isAccessCond and shouldReserve.
+// isAccessCond indicates whether the condition can be used to build ranges.
+// shouldReserve indicates whether the condition should be reserved in filter conditions.
+func (c *conditionChecker) check(condition expression.Expression) (isAccessCond, shouldReserve bool) {
 	switch x := condition.(type) {
 	case *expression.ScalarFunction:
 		return c.checkScalarFunction(x)
 	case *expression.Column:
 		if x.RetType.EvalType() == types.ETString {
-			return false
+			return false, true
 		}
 		return c.checkColumn(x)
 	case *expression.Constant:
-		return true
+		return true, false
 	}
-	return false
+	return false, true
 }
 
-func (c *conditionChecker) checkScalarFunction(scalar *expression.ScalarFunction) bool {
+func (c *conditionChecker) checkScalarFunction(scalar *expression.ScalarFunction) (isAccessCond, shouldReserve bool) {
 	_, collation := scalar.CharsetAndCollation()
 	switch scalar.FuncName.L {
 	case ast.LogicOr, ast.LogicAnd:
-		return c.check(scalar.GetArgs()[0]) && c.check(scalar.GetArgs()[1])
+		isAccessCond0, shouldReserve0 := c.check(scalar.GetArgs()[0])
+		isAccessCond1, shouldReserve1 := c.check(scalar.GetArgs()[1])
+		if isAccessCond0 && isAccessCond1 {
+			return true, shouldReserve0 || shouldReserve1
+		}
+		return false, true
 	case ast.EQ, ast.NE, ast.GE, ast.GT, ast.LE, ast.LT, ast.NullEQ:
 		if _, ok := scalar.GetArgs()[0].(*expression.Constant); ok {
-			if c.checkColumn(scalar.GetArgs()[1]) {
+			if c.matchColumn(scalar.GetArgs()[1]) {
 				// Checks whether the scalar function is calculated use the collation compatible with the column.
 				if scalar.GetArgs()[1].GetType().EvalType() == types.ETString && !collate.CompatibleCollate(scalar.GetArgs()[1].GetType().GetCollate(), collation) {
-					return false
+					return false, true
 				}
-				return scalar.FuncName.L != ast.NE || c.length == types.UnspecifiedLength
+				isFullLength := c.isFullLengthColumn()
+				if scalar.FuncName.L == ast.NE {
+					return isFullLength, !isFullLength
+				}
+				return true, !isFullLength
 			}
 		}
 		if _, ok := scalar.GetArgs()[1].(*expression.Constant); ok {
-			if c.checkColumn(scalar.GetArgs()[0]) {
+			if c.matchColumn(scalar.GetArgs()[0]) {
 				// Checks whether the scalar function is calculated use the collation compatible with the column.
 				if scalar.GetArgs()[0].GetType().EvalType() == types.ETString && !collate.CompatibleCollate(scalar.GetArgs()[0].GetType().GetCollate(), collation) {
-					return false
+					return false, true
+				}
+				isFullLength := c.isFullLengthColumn()
+				if scalar.FuncName.L == ast.NE {
+					return isFullLength, !isFullLength
```

Comment:
should this be theoretical possible
a(2) != xxx
[-min, xxx(prefix)0000 ), [xx(x+1)(prefix)FFFF, +max]

Reply:
It is theoretical possible for some types and collations. But there may be some problems like https://github.com/pingcap/tidb/issues/21145#issuecomment-1286696348 for some string types and collations.